tuned/plugins/plugin_irq.py

```diff
@@ -0,0 +1,269 @@
+from . import hotplug
+from .decorators import *
+import tuned.consts as consts
+import tuned.logs
+
+import errno
+import os
+
+log = tuned.logs.get()
+
+# the plugin manages each IRQ as a "device" and keeps a IrqInfo object for it
+class IrqInfo(object):
+	def __init__(self, irq):
+		self.irq = irq
+		self.device = "irq%s" % irq
+		self.unchangeable = False
+		self.original_affinity = None
+
+class IrqPlugin(hotplug.Plugin):
+	r"""
+	`irq`::
+
+	Allows tuning of IRQ affinities, and thus re-implements functionality
+	already present in the `scheduler` plugin. However, this plugin offers
+	more flexibility, as it allows tuning of individual interrupts with
+	different affinities. When using the `irq` plugin, make sure to disable
+	IRQ processing in the `scheduler` plugin by setting its option
+	[option]`irq_process=false`.
+	The plugin handles individual IRQs as `devices`, and multiple plugin
+	instances can be defined, each addressing different devices/irqs.
+	The `device` names used by the plugin are `irq<n>`, where `<n>` is the
+	IRQ number. The special `device` `DEFAULT` controls values written to
+	`/proc/irq/default_smp_affinity`, which applies to all non-active IRQs.
+	===
+	The option [option]`affinity` controls the IRQ affinity to be set. It is
+	a string in "cpulist" format (such as 1,3-4). If the configured affinity
+	is empty, then the affinity of the respective IRQs is not touched.
+	===
+	The option [option]`mode` is a string which can either be `set` (default)
+	or `intersect`. In `set` mode the [option]`affinity` is always written
+	as configured, whereas in `intersect` mode, the new affinity will be
+	calculated as the intersection of the current and the configured affinity.
+	If that intersection is empty, the configured affinity will be used.
+	--
+	.Example moving all IRQs to CPU0, except irq16, which is directed to CPU2
+	====
+	----
+	[irq_special]
+	type=irq
+	devices=irq16
+	affinity=2
+	[irq]
+	affinity=0
+	----
+	====
+	"""
+
+	def __init__(self, monitor_repository, storage_factory, hardware_inventory, device_matcher, device_matcher_udev, plugin_instance_factory, global_cfg, variables):
+		super(IrqPlugin, self).__init__(monitor_repository, storage_factory, hardware_inventory, device_matcher, device_matcher_udev, plugin_instance_factory, global_cfg, variables)
+		self._irqs = {}
+
+	#
+	# plugin-level methods: devices and plugin options
+	#
+	def _init_devices(self):
+		"""Read /proc/irq to collect devices
+		"""
+		self._devices_supported = True
+		self._free_devices = set()
+		self._assigned_devices = set()
+		for i in os.listdir("/proc/irq"):
+			p = os.path.join("/proc/irq", i)
+			if os.path.isdir(p) and i.isdigit():
+				info = IrqInfo(i)
+				self._irqs[i] = info
+				self._free_devices.add(info.device)
+		default_info = IrqInfo("DEFAULT")
+		default_info.device = "DEFAULT"
+		self._irqs["DEFAULT"] = default_info
+		self._free_devices.add("DEFAULT")
+
+	def _get_device_instance(self, device):
```

Comment:
What about the following:
```python
	@command_custom("mode", per_device = False, priority = -10)
	def _mode(self, enabling, value, verify, ignore_missing):
		if verify:
			return None
		if enabling and value is not None:
			self._mode_val = value
```
Then you could use the `self._mode_val` instead of the `instance._mode` and you could omit the whole `get_device_instance` method. The resulting code will be not only cleaner but will probably also run a bit faster, because it wouldn't need to scan devices to find the current instance and query it for mode.

Reply:
I agree the engine could do it better, e.g. by passing the link to instance in the `command_custom` args, but it's for future engine RFEs.

Reply:
I also don't like the `_get_device_instance`, and your suggestion solves that nicely. However, I'm also including the `verify` case, as this also needs to know the mode.